boa/src/builtins/string/mod.rs

```diff
@@ -331,11 +331,17 @@ impl String {
         // Then we convert it into a Rust String by wrapping it in from_value
         let primitive_val = this.to_string(ctx)?;
 
-        // TODO: Should throw TypeError if pattern is regular expression
-        let search_string = args
+        let arg = args
             .get(0)
-            .expect("failed to get argument for String method")
-            .to_string(ctx)?;
+            .expect("failed to get argument for String method");
```

Comment:
its better if we move this past the regexp check with the string conversion:

```
		if Self::is_regex_string(arg) {
            ctx.throw_type_error(
                "First argument to String.prototype.startsWith must not be a regular expression",
            )?;
        }

        let search_string = args.get(0).cloned().unwrap_or_else(Value::undefined).to_string(ctx)?;
```
also we shouldn't panic if the argument was not provided but we should create undefined in its place

Reply:
How about
```
        let arg = args
            .get(0)
            .cloned()
            .unwrap_or_else(Value::undefined);

        if Self::is_regexp_object(&arg) {
            ctx.throw_type_error(
                "First argument to String.prototype.startsWith must not be a regular expression",
            )?;
        }

        let search_string = arg.to_string(ctx)?;
```

`is_regexp_object` takes `&Value` so we'd need to `get(0)` before calling it.